realtek: add reset button and reboot fix for D-Link DGS-1210-28

luizluca opened the pull request:
Tested in a DGS-1210-28 F3, both triggering failsafe and reboot.

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

luizluca committed:
realtek: add reset button for D-Link DGS-1210-28

Tested in a DGS-1210-28 F3, both triggering failsafe and reboot.

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

luizluca commented:
BTW, D-Link DGS-1210-28 is freezing during a reboot after "reboot: Restarting system". It change the led frequency and that's it. It never gets to the bootloader again. I'll take a look.

robimarko commented:
Regarding the reset, it's a known issue that can be fixed by using a GPIO for restart instead.
I have been tracing the SFP-s and rest of the stuff and queing for submission here:
https://github.com/sartura/openwrt/tree/realtek/dgs-1210-28p

luizluca commented:
> Regarding the reset, it's a known issue that can be fixed by using a GPIO for restart instead. I have been tracing the SFP-s and rest of the stuff and queing for submission here: https://github.com/sartura/openwrt/tree/realtek/dgs-1210-28p

Thanks @robimarko. I got the reset-gpio solution a couple of minutes before your reply. I'll not submit it then. Anyway, the reset button is not in that tree. We could either add it there or merge it independently.

robimarko commented:
It's probably the fastest to get it merged via this PR.
Do you know if it's common for all of the DGS-1210 series like LED-s are?

luizluca committed:
realtek: add gpio-restart for D-Link DGS-1210-28

A GPIO assert is required to reset the system. Otherwise, the system
will hang on reboot.

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

musashino205 commented on target/linux/realtek/image/rtl838x.mk:
```diff
@@ -33,6 +33,7 @@ TARGET_DEVICES += d-link_dgs-1210-16
 define Device/d-link_dgs-1210-28
   $(Device/d-link_dgs-1210)
   DEVICE_MODEL := DGS-1210-28
+  DEVICE_PACKAGES += kmod-input-gpio-keys-polled
```
The same feature is provided by **kmod-gpio-button-hotplug** selected by [```realtek/Makefile```](https://github.com/openwrt/openwrt/blob/1521d5f453683ca1aae87d025d2892901369ce0c/target/linux/realtek/Makefile#L20), so this line can be dropped.

luizluca replied:
Thanks! Updated

luizluca replied:
Sorry, my bad. Done.

robimarko commented:
@luizluca Sure, I can make a PR for the reboot GPIO as it's vital.

robimarko commented on target/linux/realtek/dts-5.10/rtl8382_d-link_dgs-1210-28.dts:
```diff
@@ -5,6 +5,24 @@
+		};
+	};
+
+	gpio1: rtl8231-gpio {
```
Can we stop using the gpio1 label?
It doesn't really make sense, calling it just rtl8231 would be probably better as this way it looks like its just another GPIO controller from the CPU.

luizluca replied:
Whatever we call it, we should keep it the same for every device. Either we rename all existing references before this PR or we keep gpio1 here as is and rename all of them afterwards.

robimarko commented:
Drop `kmod-input-gpio-keys-polled`, and after that:
`Reviewed-by: Robert Marko <robimarko@gmail.com>`

svanheule commented:
If the systems isn't rebooting properly, try adding the following override to the device DTS:
```yaml
&watchdog0 {
    realtek,reset-mode = "cpu";
};
```

This makes sure the reset doesn't touch the SPI peripheral, which is what's causing this hang IIRC. Using a value of `"software"` is also possible, but that doesn't reset any hardware, so should only be used as a last resort.

As mentioned here, `gpio-restart` is indeed the preferred reboot method, but I think you should have both (and test them individually). Even if it's just to document that this is one of those devices that requires special care to restart.

luizluca commented:
> ```yaml
> &watchdog0 {
>     realtek,reset-mode = "cpu";
> };
> ```

If I didn't do anything wrong, "cpu" and "software" didn't work.

```
# cat /proc/device-tree/soc/watchdog@3150/realtek,reset-mode; echo
cpu
```

```
# cat /proc/device-tree/soc/watchdog@3150/realtek,reset-mode; echo
software
```

With either cpu or software and gpio-reset disabled, I still get locked in a state where the system led blink and sporadically ports 26 and 27 blinks, one in green and the other in amber. Sometimes, I get ports 26 and 27 always on. Does it mean that the watchdog is not able to reboot the system in case it freezes? Can't watchdog use the gpio?

robimarko commented:
They won't work, I have tried that way back when I got the switch, only GPIO-based restart works on DGS.

luizluca commented:
> They won't work, I have tried that way back when I got the switch, only GPIO-based restart works on DGS.

So, the patch is as good as we can get.

Is there any method to install OpenWrt that does not use the serial cable? I tried to trigger some form of emergency firmware page or even a non-interactive TFTP but nothing seems to work. And there is no factory image. The web gui rejects both initramfs or sysupgrade. And the serial port is not easily accessible.

The [commit message](b5bd945733a7ed235777c2ceaa0ff27dc779630f) from @ynezz that added the support did mentioned to flash from OEM webpage.

"To install, upload the sysupgrade image to the OEM webpage or sysupgrade
from the system running from initramfs image."

I guess "OEM" here really means "Luci" after initramfs is already in use.

robimarko commented:
Even if there was an image that stock FW WEB UI will accept, the stock kernel partition size is 1.X MB only.
So flashing would fail anyway.

robimarko commented:
Yeah, the FW image is basically just a container for the kernel and squashfs with a header.
It uses the header to figure out whats kernel and what squashfs, plus there is a checksum for both of them, and then writes them to the selected kernel and rootfs partition (1 or 2 depending on the target one you choose) and that's it.

In theory its possible, but there is no way to generate a kernel like that in OpenWrt, nor its worth adding quirks for every device supported.

xabolcs commented:
> Wouldn't it be possible to ship a slim <1M kernel that only kexec a full featured /boot/vmlinuz kernel?

> In theory its possible, but there is no way to generate a kernel like that in OpenWrt, nor its worth adding quirks for every device supported.

Hi @luizluca,

as @robimarko wrote, it's doable, but you have to decide it's worth your time or not!
For example **blocktrron** did some magic with RAVPower RP-WD009 in commit e959048c128e4350b78df653be7cd7277787518e:
* he ported the device to upstream U-Boot, then he
* wrote the upstream U-Boot to the 1.5 MByte sized `Kernel_RootFS` vendor partition
* wrote OpenWrt kernel + OpenWrt rootfs to `Rootfs` vendor partition

And when the device booted ...:
* the vendor U-Boot booted the upstream U-Boot
* upstream U-Boot booted the OpenWrt `firmware` (kernel+rootfs)

svanheule commented:
Merged to master and I'll apply the fixes to 22.03 too. Feel free to continue the firmware discussion on the forum, so this PR isn't too polluted with other discussions.

luizluca commented:
> > Wouldn't it be possible to ship a slim <1M kernel that only kexec a full featured /boot/vmlinuz kernel?
> 
> > In theory its possible, but there is no way to generate a kernel like that in OpenWrt, nor its worth adding quirks for every device supported.
> 
> Hi @luizluca,
> 
> as @robimarko wrote, it's doable, but you have to decide it's worth your time or not! For example **blocktrron** did some magic with RAVPower RP-WD009 in commit [e959048](https://github.com/openwrt/openwrt/commit/e959048c128e4350b78df653be7cd7277787518e):
> 
> * he ported the device to upstream U-Boot, then he
> * wrote the upstream U-Boot to the 1.5 MByte sized `Kernel_RootFS` vendor partition
> * wrote OpenWrt kernel + OpenWrt rootfs to `Rootfs` vendor partition
> 
> And when the device booted ...:
> 
> * the vendor U-Boot booted the upstream U-Boot
> * upstream U-Boot booted the OpenWrt `firmware` (kernel+rootfs)

Thanks @xabolcs, I might take a look on this. It looks like a nice approach.

I might need to install OpenWrt about 150 switches. Maybe porting uboot might to be easier than opening all those cases and manually holding some wires ;-). Leaving the thread.

Thanks @svanheule

luizluca commented:
> Hi @luizluca,
> 
> as @robimarko wrote, it's doable, but you have to decide it's worth your time or not!

I did it. It was worth it. #10113
